projects/validation-messages/src/lib/components/validation-messages/validation-messages.component.ts

```diff
@@ -142,74 +116,105 @@ export class ValidationMessagesComponent
-    if (
-      this.control &&
-      ((this.control.invalid && this.control.touched) ||
-        (!this.control.invalid && this.shownErrors.length > 0))
-    ) {
-      this.updateErrorMessages();
-      this.cd.markForCheck();
-    }
-  }
-
-  private updateErrorMessages(): void {
-    this.shownErrors = [];
+  private _updateErrorMessages(): void {
+    this.shownErrors.set([]);
 
     if (!this.control || !this.control.errors) {
       return;
     }
 
     const controlErrors = this.control.errors;
+    let errorCounter = 0;
     for (const propertyName in controlErrors) {
-      if (!this.multiple && this.shownErrors.length === 1) {
+      if (!this._multiple() && errorCounter === 1) {
         break;
       }
 
       if (controlErrors[propertyName] && propertyName !== 'server') {
-        this.shownErrors.push(
+        const errorMessage =
           this.validationMessagesService.getValidatorErrorMessage(
             propertyName,
             controlErrors[propertyName],
             this.errorsMessages
-          )
-        );
+          );
+
+        this.shownErrors.update((errors) => [...errors, errorMessage]);
+        errorCounter++;
       }
     }
   }
+  private _valueChangesEffect() {
+    effect(
+      () => {
+        this.controlValueChanges();
+
+        if (
+          this.control &&
+          ((this.control.invalid && this.control.dirty) ||
+            (!this.control.invalid && this.shownErrors.length > 0))
+        ) {
+          this._updateErrorMessages();
+        }
+      },
+      {
+        allowSignalWrites: true,
+      }
+    );
+  }
+
+  private _apiErrorMessagesEffect() {
+    effect(() => {
+      this._clearErrors();
+      const errorMessages = this._apiErrorMessages();
+      this.parseApiErrorMessages(errorMessages);
+      this.showServerErrors = true;
+
+      if (this.control && errorMessages) {
+        this.control.setErrors({
+          server: errorMessages,
+        });
+      }
+    });
+  }
 
-  private unsubscribeAndClearValueChanges(): void {
+  private _multipleEffect() {
+    effect(
+      () => {
+        this._multiple();
+        this._updateErrorMessages();
+      },
+      {
+        allowSignalWrites: true,
+      }
+    );
+  }
+
+  private _clearErrors(): void {
     if (this.control) {
       this.control.setErrors({});
       this.control.updateValueAndValidity({ onlySelf: true, emitEvent: false });
     }
+    this.showServerErrors = false;
+  }
 
-    if (this.valueChanges && !this.valueChanges.closed) {
-      this.valueChanges.unsubscribe();
-    }
+  private _listenToFormChanges(): void {
+    const valueChanges$ = this.control.valueChanges;
 
-    this.showServerErrors = false;
-    this.valueChanges = null;
+    valueChanges$.pipe(takeUntilDestroyed(this.destroyRef)).subscribe(() => {
+      this._clearErrors();
+    });
+
+    this.controlValueChanges = toSignal(valueChanges$, {
+      injector: this.injector,
+      initialValue: null,
+    });
```

Comment:
is it necessary if  `valueChanges$` is already subscribed? you can keep the logic there in one place, without spreading it to multiple places like `_valueChangesEffect` method

Reply:
or just handle it within the effect